File: bin/script/Component/MysqlProxyComponent.lua
local MysqlProxyComponent = { }

local type = type
local node = require("Node")
local json = require("util.json")
local table_insert = table.insert
local table_concat = table.concat
local string_format = string.format

local MYSQL_READ_PROXY = "MysqlReadProxy"
local MYSQL_WRITE_PROXY = "MysqlWriteProxy"

---@param tab string
---@param keys table
function MysqlProxyComponent:Create(tab, keys)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code = node:Call(id, "MysqlWriteProxy.Create", {
        tab = tab,
        keys = keys
    })
    return code == XCode.Ok
end

---@param tab string
---@param field string
---@param value number
---@param filter table
function MysqlProxyComponent:Inc(tab, filter, field, value)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code, response = node:Call(id, "MysqlWriteProxy.Inc", {
        tab = tab,
        field = field,
        value = value or 1,
        filter = filter
    })
    return code == XCode.Ok and response.value or 0
end

---@param list table
function MysqlProxyComponent:Commit(list)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code, response = node:Call(id, "MysqlWriteProxy.Commit", list)
    return code == XCode.Ok and response or nil
end

---@param tab string
---@param document table
---@return number
function MysqlProxyComponent:InsertOne(tab, document)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code, response = node:Call(id, "MysqlWriteProxy.InsertOne", {
        tab = tab,
        document = document
    })
    if code ~= XCode.Ok then
        return 0
    end
    return response.count
end

---@param tab string
---@param documents table
---@return number
function MysqlProxyComponent:InsertBatch(tab, documents)
    assert(#documents > 0)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code, response = node:Call(id, "MysqlWriteProxy.InsertBatch", {
        tab = tab,
        documents = documents
    })
    if code ~= XCode.Ok then
        return 0
    end
    return response.count
end

---@param tab string
---@param document table
---@return boolean
function MysqlProxyComponent:ReplaceOne(tab, document)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code = node:Call(id, "MysqlWriteProxy.Replace", {
        tab = tab,
        document = document
    })
    return code == XCode.Ok
end

---@param tab string
---@param filter table | string
---@return boolean
function MysqlProxyComponent:Delete(tab, filter, limit)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code = node:Call(id, "MysqlWriteProxy.Delete", {
        tab = tab,
        limit = limit or 1,
        filter = filter
    })
    return code == XCode.Ok
end

---@param tab string
---@param filter table | string
---@param document table
---@param limit number
---@return boolean
function MysqlProxyComponent:Update(tab, filter, document, limit)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code, response = node:Call(id, "MysqlWriteProxy.Update", {
        tab = tab,
        limit = limit or 1,
        filter = filter,
        document = document
    })
    return code == XCode.Ok and response.count, 0
end

---@param tab string
function MysqlProxyComponent:Drop(tab)
    return self:RunInWrite(string_format("DROP TABLE %s;", tab))
end

---@param tab string
---@param field string
---@param unique boolean
---@return boolean
function MysqlProxyComponent:SetIndex(tab, field, unique)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code = node:Call(id, "MysqlWriteProxy.SetIndex", {
        tab = tab,
        field = field,
        unique = unique or false
    })
    return code == XCode.Ok
end

---@param tab string
---@param filter table | string
---@param fields table
---@return table | nil
function MysqlProxyComponent:FindOne(tab, filter, fields)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.FindOne", {
        tab = tab,
        limit = 1,
        filter = filter,
        fields = fields,
    })
    return code ~= XCode.Ok and nil or response
end

---@param tab string
---@param filter table | string
---@param fields table
---@return table | nil
function MysqlProxyComponent:Find(tab, filter, fields, sort, limit)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.Find", {
        tab = tab,
        sort = sort,
        limit = limit,
        filter = filter,
        fields = fields,
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response
end

---@param tab string
---@param field string
---@param value string
---@param fields table
---@param limit number
---@return table | nil
function MysqlProxyComponent:Like(tab, field, value, fields, sort, limit)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.Like", {
        tab = tab,
        sort = sort,
        field = field,
        value = value,
        fields = fields,
        limit = limit or 1
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response
end

---@param stmt string
---@param args table
function MysqlProxyComponent:ExecuteInRead(stmt, args)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.Execute", {
        stmt = stmt,
        args = args
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response
end

---@param tab string
---@param func string
---@param field string
---@param filter table
---@return table
function MysqlProxyComponent:Func(tab, func, field, filter, group)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.Func", {
        tab = tab,
        func = func,
        field = field,
        filter = filter,
        group =group,
    })
    return code ~= XCode.Ok and nil or response
end

---@param tab string
---@param filter table
---@return number
function MysqlProxyComponent:Count(tab, filter)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.Count", {
        tab = tab,
        filter = filter
    })
    if code ~= XCode.Ok then
        return 0
    end
    return response.count
end

---@param tab string
---@param filter table | string
---@param fields table
---@param page number
---@param count number
---@return table | nil
function MysqlProxyComponent:FindPage(tab, filter, sort, fields, page, count)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.FindPage", {
        tab = tab,
        sort = sort,
        filter = filter,
        fields = fields,
        page = page or 1,
        count = count or 10
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response.list
end

---@param sql string
function MysqlProxyComponent:RunInRead(sql)
    local id = node:Allot(MYSQL_READ_PROXY)
    local code, response = node:Call(id, "MysqlReadProxy.Run", sql)
    if code ~= XCode.Ok then
        return nil
    end
    return response.list, response.count
end

---@param sql string
function MysqlProxyComponent:RunInWrite(sql)
    local id = node:Allot(MYSQL_WRITE_PROXY)
    local code, response = node:Call(id, "MysqlWriteProxy.Run", sql)
    if code ~= XCode.Ok then
        return nil
    end
    return response.list, response.count
end

return MysqlProxyComponent

File: Common/DB/Mysql/Component/MysqlProxyComponent.cpp
//
// Created by yy on 2025/7/6.
//

#include "MysqlProxyComponent.h"
#include "Mysql/Service/MysqlReadProxy.h"
#include "Mysql/Service/MysqlWriteProxy.h"
#include "Node/Component/NodeComponent.h"
#include "Entity/Component/ComponentFactory.h"
namespace acs
{
    MysqlProxyComponent::MysqlProxyComponent()
    {
        this->mNode = nullptr;
        this->mReadName = ComponentFactory::GetName<MysqlReadProxy>();
        this->mWriteName = ComponentFactory::GetName<MysqlWriteProxy>();
    }

    bool MysqlProxyComponent::LateAwake()
    {
        this->mNode = this->GetComponent<NodeComponent>();
        return true;
    }

    int MysqlProxyComponent::InsertOne(const char* tab, const pb::Message& message)
    {
        std::string document;
        if(!pb_json::MessageToJsonString(message, &document).ok())
        {
            return XCode::ProtoCastJsonFailure;
        }
        json::w::Document request;
        request.Add("tab", tab);
        request.AddObject("document", document);
        const std::string func = this->mWriteName + ".InsertOne";
        return this->CallWriteProxy(func, request);
    }

    int MysqlProxyComponent::InsertOne(const char* tab, const json::w::Document& document)
    {
        json::w::Document request;
        request.Add("tab", tab);
        request.Add("document", document);
        const std::string func = this->mWriteName + ".InsertOne";
        return this->CallWriteProxy(func, request);
    }

    int MysqlProxyComponent::ReplaceOne(const char* tab, const json::w::Document& document)
    {
        json::w::Document request;
        request.Add("tab", tab);
        request.Add("document", document);
        const std::string func = this->mWriteName + ".Replace";
        return this->CallWriteProxy(func, request);
    }

    int MysqlProxyComponent::DeleteOne(const char* tab, const json::w::Document& filter)
    {
        json::w::Document request;
        {
            request.Add("tab", tab);
            request.Add("limit", 1);
            request.Add("filter", filter);
        }
        const std::string func = this->mWriteName + ".Delete";
        return this->CallWriteProxy(func, request);
    }

    int MysqlProxyComponent::UpdateOne(const char* tab, const json::w::Document& filter, const json::w::Document& document)
    {
        json::w::Document request;
        {
            request.Add("tab", tab);
            request.Add("limit", 1);
            request.Add("filter", filter);
            request.Add("document", document);
        }
        const std::string func = this->mWriteName + ".Update";
        return this->CallWriteProxy(func, request);
    }

    int MysqlProxyComponent::CallWriteProxy(const std::string& func, const json::w::Document& request)
    {
        Node * mysqlProxy = this->mNode->Next(this->mWriteName);
        if(mysqlProxy == nullptr)
        {
            return XCode::NotFoundActor;
        }
        std::unique_ptr<json::r::Document> response = std::make_unique<json::r::Document>();
        if( mysqlProxy->Call(func, request, response) != XCode::Ok)
        {
            return XCode::Failure;
        }
        int count = 0;
        return response->Get("count", count) && count >= 1 ? XCode::Ok : XCode::Failure;
    }

    int MysqlProxyComponent::RunInRead(const std::string& sql, std::unique_ptr<json::r::Document>& response)
    {
        Node * mysqlProxy = this->mNode->Next(this->mReadName);
        if(mysqlProxy == nullptr)
        {
            return XCode::NotFoundActor;
        }
        const std::string func = this->mReadName + ".Run";
        return mysqlProxy->Call(func, sql, response);
    }


    int MysqlProxyComponent::RunInWrite(const std::string& sql, std::unique_ptr<json::r::Document>& response)
    {
        Node * mysqlProxy = this->mNode->Next(this->mWriteName);
        if(mysqlProxy == nullptr)
        {
            return XCode::NotFoundActor;
        }
        const std::string func = this->mWriteName + ".Run";
        return mysqlProxy->Call(func, sql, response);
    }

    long long MysqlProxyComponent::Inc(const char* tab, const char* field, const json::w::Document& filter, int value)
    {
        long long result = 0;
        do
        {
            Node * mysqlProxy = this->mNode->Next(this->mWriteName);
            if(mysqlProxy == nullptr)
            {
                result = -1;
                break;
            }
            json::w::Document request;
            {
                request.Add("tab", tab);
                request.Add("field", field);
                request.Add("value", value);
                request.Add("filter", filter);
            }
            const std::string func = this->mWriteName + ".Inc";
            std::unique_ptr<json::r::Document> response = std::make_unique<json::r::Document>();
            if(mysqlProxy->Call(func, request, response) != XCode::Ok)
            {
                result = 0;
            }
            response->Get("value", result);
        }
        while(false);
        return result;
    }

    int MysqlProxyComponent::FindOne(const char* tab, const json::w::Document& filter, std::unique_ptr<json::r::Document>& document)
    {
        Node * mysqlProxy = this->mNode->Next(this->mWriteName);
        if(mysqlProxy == nullptr)
        {
            return  XCode::NotFoundActor;
        }
        json::w::Document request;
        {
            request.Add("tab", tab);
            request.Add("filter", filter);
        }
        const std::string func = this->mWriteName + ".FindOne";
        return  mysqlProxy->Call(func, request, document);
    }

    int MysqlProxyComponent::Find(const char* tab, const json::w::Document& filter, std::unique_ptr<json::r::Document>& document)
    {
        Node * mysqlProxy = this->mNode->Next(this->mWriteName);
        if(mysqlProxy == nullptr)
        {
            return  XCode::NotFoundActor;
        }
        json::w::Document request;
        {
            request.Add("tab", tab);
            request.Add("filter", filter);
        }
        const std::string func = this->mWriteName + ".Find";
        return  mysqlProxy->Call(func, request, document);
    }

    int MysqlProxyComponent::Find(const char* tab, const std::list<std::string>& fields, const json::w::Document& filter, std::unique_ptr<json::r::Document>& document)
    {
        Node * mysqlProxy = this->mNode->Next(this->mWriteName);
        if(mysqlProxy == nullptr)
        {
            return  XCode::NotFoundActor;
        }
        json::w::Document request;
        {
            request.Add("tab", tab);
            request.Add("filter", filter);
            request.Add("fields", fields);
        }
        const std::string func = this->mWriteName + ".Find";
        return  mysqlProxy->Call(func, request, document);
    }

    int MysqlProxyComponent::FindOne(const char* tab, const std::list<std::string>& fields, const json::w::Document& filter, std::unique_ptr<json::r::Document>& document)
    {
        Node * mysqlProxy = this->mNode->Next(this->mReadName);
        if(mysqlProxy == nullptr)
        {
            return  XCode::NotFoundActor;
        }
        json::w::Document request;
        {
            request.Add("tab", tab);
            request.Add("filter", filter);
            request.Add("fields", fields);
        }
        const std::string func = this->mReadName + ".FindOne";
        return  mysqlProxy->Call(func, request, document);
    }

    long long MysqlProxyComponent::Count(const char* tab, const json::w::Document& filter)
    {
        long long result = 0;
        do
        {
            Node * mysqlProxy = this->mNode->Next(this->mReadName);
            if(mysqlProxy == nullptr)
            {
                result = -1;
                break;
            }
            json::w::Document request;
            {
                request.Add("tab", tab);
                request.Add("filter", filter);
            }
            const std::string func = this->mReadName + ".Count";
            std::unique_ptr<json::r::Document> response = std::make_unique<json::r::Document>();
            if(mysqlProxy->Call(func, request, response) != XCode::Ok)
            {
                result = 0;
                break;
            }
            response->Get("count", result);
        }
        while(false);
        return result;
    }
}